### PR TITLE
Feature/batch meta

### PR DESCRIPTION
**Pull request by Solonets**
### Description of the Change

 Update transaction protobuf with new batch payload field:epts. -->

### Benefits

Needed for the batch processing

### Possible Drawbacks 

--

### CODE REVIEW

File: shared_model/builders/protobuf/builder_templates/transaction_template.hpp [hunk 1]
```diff
@@ -105,19 +107,23 @@ namespace shared_model {
...
         });
       }
 
       auto quorum(interface::types::QuorumType quorum) const {
-        return transform<Quorum>(
-            [&](auto &tx) { tx.mutable_payload()->set_quorum(quorum); });
+        return transform<Quorum>([&](auto &tx) {
+          tx.mutable_payload()->mutable_reduced_payload()->set_quorum(quorum);
+        });
       }
```
Comment (kamilsa) [inline shared_model/builders/protobuf/builder_templates/transaction_template.hpp]: Please add batch meta field setting to the builder

File: shared_model/interfaces/common_objects/types.hpp [hunk 2]
```diff
@@ -98,6 +98,8 @@ namespace shared_model {
                            const AccountAsset &>;
       /// Type of the transfer message
       using DescriptionType = std::string;
+
+      using BatchType = enum { ATOMIC, ORDERED };
```
Comment (kamilsa) [inline shared_model/interfaces/common_objects/types.hpp]: Please use enum class

File: shared_model/backend/protobuf/batch_meta.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
+/**
```
Comment (kamilsa) [inline shared_model/backend/protobuf/batch_meta.hpp]: Use new header please

File: shared_model/backend/protobuf/transaction.hpp [hunk 4]
```diff
@@ -83,11 +88,20 @@ namespace shared_model {
...
       interface::types::TimestampType createdTime() const override {
-        return payload_.created_time();
+        return reduced_payload_.created_time();
       }
 
       interface::types::QuorumType quorum() const override {
-        return payload_.quorum();
+        return reduced_payload_.quorum();
+      }
+      boost::optional<std::shared_ptr<interface::BatchMeta>> batch_meta()
```
Comment (Solonets) [inline shared_model/backend/protobuf/transaction.hpp]: ?

File: shared_model/backend/protobuf/batch_meta.hpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+#include "interfaces/common_objects/amount.hpp"
+
+#include <numeric>
+
+#include "backend/protobuf/common_objects/trivial_proto.hpp"
+#include "backend/protobuf/util.hpp"
+#include "block.pb.h"
+#include "interfaces/common_objects/types.hpp"
+#include "utils/lazy_initializer.hpp"
+#include <boost/range/numeric.hpp>
```
Comment (victordrobny) [inline shared_model/backend/protobuf/batch_meta.hpp]: Please, reorder includes according to google code style

File: shared_model/backend/protobuf/transaction.hpp [hunk 6]
```diff
@@ -96,12 +96,7 @@ namespace shared_model {
       }
```
Comment (kamilsa) [inline shared_model/backend/protobuf/transaction.hpp]: add blank line

File: shared_model/builders/protobuf/builder_templates/transaction_template.hpp [hunk 7]
```diff
@@ -105,19 +107,35 @@ namespace shared_model {
...
+          tx.mutable_payload()->mutable_batch()->set_type(
+              static_cast<
+                  iroha::protocol::Transaction::Payload::BatchMeta::BatchType>(
+                  type));
+          for (auto hash : hashes) {
+            tx.mutable_payload()->mutable_batch()->add_tx_hashes(
+                hash.toString());
+          }
         });
       }
```
Comment (kamilsa) [inline shared_model/builders/protobuf/builder_templates/transaction_template.hpp]: Return blank line

File: shared_model/builders/protobuf/builder_templates/transaction_template.hpp [hunk 8]
```diff
@@ -105,19 +107,35 @@ namespace shared_model {
...
+        });
+      }
+      auto batchMeta(interface::types::BatchType type,
+                     std::vector<interface::types::HashType> hashes) const {
+        return transform<0>([&](auto &tx) {
+          tx.mutable_payload()->mutable_batch()->set_type(
+              static_cast<
+                  iroha::protocol::Transaction::Payload::BatchMeta::BatchType>(
+                  type));
+          for (auto hash : hashes) {
```
Comment (kamilsa) [inline shared_model/builders/protobuf/builder_templates/transaction_template.hpp]: Can be const auto&

Comment (victordrobny) [inline shared_model/builders/protobuf/builder_templates/transaction_template.hpp]: Please, use &

File: shared_model/builders/protobuf/builder_templates/transaction_template.hpp [hunk 9]
```diff
@@ -105,19 +107,35 @@ namespace shared_model {
...
+            tx.mutable_payload()->mutable_batch()->add_tx_hashes(
+                hash.toString());
```
Comment (kamilsa) [inline shared_model/builders/protobuf/builder_templates/transaction_template.hpp]: `toString` should be used only for debugging purpose, please use `crypto::toBinaryString(hash))` instead

File: shared_model/interfaces/transaction.hpp [hunk 10]
```diff
@@ -52,6 +54,23 @@ namespace shared_model {
...
+       */
+      virtual const types::BlobType &reduced_payload() const = 0;
+
+      const types::HashType &reduced_hash() const {
+        if (reduced_hash_ == boost::none) {
+          reduced_hash_.emplace(HashProvider::makeHash(reduced_payload()));
+        }
+        return *reduced_hash_;
+      }
+      /*
```
Comment (kamilsa) [inline shared_model/interfaces/transaction.hpp]: Please separate new methods with blank lines

**Pull request merged**